pkg/kubeapiserver/options/authentication.go

```diff
@@ -654,18 +664,27 @@ func (o *BuiltInAuthenticationOptions) ApplyTo(ctx context.Context, authInfo *ge
 	authInfo.Authenticator = authenticator
 
 	if len(o.AuthenticationConfigFile) > 0 {
+		authenticationconfigmetrics.RegisterMetrics()
 		trackedAuthenticationConfigData := authenticatorConfig.AuthenticationConfigData
 		var mu sync.Mutex
 		go filesystem.WatchUntil(
 			ctx,
 			time.Minute,
 			o.AuthenticationConfigFile,
 			func() {
-				// TODO add metrics
 				// TODO collapse onto shared logic with DynamicEncryptionConfigContent controller
+				var (
+					err        error
+					authConfig *apiserver.AuthenticationConfiguration
+				)
 
 				mu.Lock()
-				defer mu.Unlock()
+				defer func() {
+					if err != nil {
+						authenticationconfigmetrics.RecordAuthenticationConfigAutomaticReloadFailure(apiServerID)
```

Comment:
I think this would actually be easier to follow if we log and record the failure metric in each `if err != nil` block (xref reloadableAuthorizerResolver#checkFile)

Reply:
ha, I see @enj suggested a defer in https://github.com/enj/kubernetes/pull/1#discussion_r1518673816 ... relying on overwriting `err` through the whole function feels fragile and I find the defer harder to reason about, but I won't block if both if you think it's better

Reply:
Keeping it inline so we know exactly when we want to record a failure metric.